### PR TITLE
store: add logic to drop outdated blocks

### PR DESCRIPTION
**Pull request by fabxc**

### CODE REVIEW

File: pkg/store/bucket.go [hunk 1]
```diff
@@ -264,12 +264,15 @@ func (s *BucketStore) SyncBlocks(ctx context.Context) error {
 		}()
 	}
 
+	allIDs := map[ulid.ULID]struct{}{}
+
 	err := s.bucket.Iter(ctx, "", func(name string) error {
 		// Strip trailing slash indicating the a directory.
 		id, err := ulid.Parse(name[:len(name)-1])
```
Comment (Bplotka) [inline pkg/store/bucket.go]: not yours, but.. line above `s~// Strip trailing slash indicating the a directory.~// Strip trailing slash indicating a directory.~`

File: pkg/store/bucket.go [hunk 2]
```diff
@@ -284,7 +287,19 @@ func (s *BucketStore) SyncBlocks(ctx context.Context) error {
...
-	return err
+	if err != nil {
+		return err
+	}
+	// Drop all blocks that are no longer present in the bucket.
+	for id := range s.blocks {
+		if _, ok := allIDs[id]; ok {
+			continue
+		}
+		if err := s.removeBlock(id); err != nil {
```
Comment (Bplotka) [inline pkg/store/bucket.go]: wow, lot's of races here (: during compaction

Comment (Bplotka) [reply pkg/store/bucket.go]: but I guess we just end-up having maybe weird results during compaction on query level.

Comment (fabxc) [reply pkg/store/bucket.go]: Well, for the duration a new block and its sources exist concurrently, nothing weird should be showing. It's the same data and we should have measures in place to skip over those.

What could happen is that we delete old blocks before properly seeing their replacement. That should be rare and can be addressed if it occurs in practice.

**Pull request merged**